BLE Indoor Positioning/src/main/java/com/nexenio/bleindoorpositioning/ble/beacon/signal/KalmanFilter.java

```diff
@@ -0,0 +1,92 @@
+package com.nexenio.bleindoorpositioning.ble.beacon.signal;
+
+import com.nexenio.bleindoorpositioning.ble.advertising.AdvertisingPacket;
+import com.nexenio.bleindoorpositioning.ble.advertising.AdvertisingPacketUtil;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Created by leon on 02.01.18.
+ *
+ * Kalman filtering, also known as linear quadratic estimation (LQE), is an algorithm that uses a
+ * series of measurements observed over time, containing statistical noise and other inaccuracies,
+ * and produces estimates of unknown variables that tend to be more accurate than those based on a
+ * single measurement alone, by using Bayesian inference and estimating a joint probability
+ * distribution over the variables for each timeframe.
+ *
+ * <a href="https://en.wikipedia.org/wiki/Kalman_filter">Kalman Filter</a> <a
+ * href="https://www.wouterbulten.nl/blog/tech/kalman-filters-explained-removing-noise-from-rssi-signals/">Kalman
+ * Explained</a>
+ *
+ * Since RSSI signals are largely influenced by signal noise, taking samples from the signal seems
+ * likely to be beneficial. --> Evaluate Unscented Kalman filter
+ *
+ * inspired by https://github.com/fgroch/beacon-rssi-resolver/blob/master/src/main/java/tools/blocks/filter/KalmanFilter.java
+ */
+
+public class KalmanFilter implements RssiFilter {
+
+    public static final long DURATION_DEFAULT = TimeUnit.SECONDS.toMillis(3);
+
+    private long minimumTimestamp;
+    private long maximumTimestamp;
+    private double processNoise;
+    private double measurementNoise;
+    private double estimatedRSSI;
+    private double errorCovarianceRSSI;
+    private boolean isInitialized = false;
+
+    public KalmanFilter() {
+        maximumTimestamp = System.currentTimeMillis();
```

Comment:
Change to `this(System.currentTimeMillis(), maximumTimestamp - DURATION_DEFAULT)` or else `processNoise` and  `measurementNoise` will not be set if this constructor is called.